File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if(nums.size()<3){ // no answer exists
            return {};
        }
        vector<vector<int>> triplets;
        sort(nums.begin(), nums.end());
        if(nums[0] > 0 || nums.back() < 0){ // all positive or negative then no answer
            return {};
        }
        for(int firstElement = 0; firstElement < nums.size() - 2 && nums[firstElement] <= 0; ++firstElement){
            if(firstElement > 0 && nums[firstElement]==nums[firstElement-1]){ // exclude duplicates for 1st element
                continue;
            }
            int secondElement = firstElement + 1, thirdElement = nums.size() - 1;
            while(secondElement < thirdElement){ // 2nd element comes before 3rd element
                int sum = nums[firstElement] + nums[secondElement] + nums[thirdElement];
                if(sum < 0){
                    secondElement++;
                }
                else if(sum > 0){
                    thirdElement--;
                }
                else{
                    triplets.push_back({nums[firstElement],nums[secondElement],nums[thirdElement]});
                    secondElement++;
                    while(secondElement < thirdElement && nums[secondElement] == nums[secondElement -  1]){ // exclude duplicates for 2nd element
                        secondElement++;
                    }
                    thirdElement--;
                    while(secondElement < thirdElement && nums[thirdElement] == nums[thirdElement + 1]){ // exclude dupplicates for 3rd element
                        thirdElement--;
                    }
                }
            }
        }
        return triplets;
    }
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] num) {
        List<List<Integer>> res = new ArrayList<>();

        Arrays.sort(num);

        for (int i = 0; i < num.length; i++) {

            int target = -num[i];
            int front = i + 1;
            int back = num.length - 1;

            if(target < 0)
            {
                break;
            }

            while (front < back) {

                int sum = num[front] + num[back];

                // Finding answer which start from number num[i]
                if (sum < target)
                    front++;

                else if (sum > target)
                    back--;

                else {
                    List<Integer> triplet = new ArrayList<>();
                    triplet.add(num[i]);
                    triplet.add(num[front]);
                    triplet.add(num[back]);
                    res.add(triplet);

                    // Processing duplicates of Number 2
                    // Rolling the front pointer to the next different number forwards
                    while (front < back && num[front] == triplet.get(1)) front++;

                    // Processing duplicates of Number 3
                    // Rolling the back pointer to the next different number backwards
                    while (front < back && num[back] == triplet.get(2)) back--;
                }

            }

            // Processing duplicates of Number 1
            while (i + 1 < num.length && num[i + 1] == num[i]) 
                i++;

        }
        return res;

    }
}